executor/analyze.go

```diff
@@ -496,22 +527,464 @@ type AnalyzeFastTask struct {
 // AnalyzeFastExec represents Fast Analyze executor.
 type AnalyzeFastExec struct {
 	ctx             sessionctx.Context
-	PhysicalTableID int64
+	physicalTableID int64
 	pkInfo          *model.ColumnInfo
 	colsInfo        []*model.ColumnInfo
 	idxsInfo        []*model.IndexInfo
 	concurrency     int
 	maxNumBuckets   uint64
-	table           table.Table
 	cache           *tikv.RegionCache
 	wg              *sync.WaitGroup
 	sampLocs        chan *tikv.KeyLocation
-	sampLocRowCount uint64
-	tasks           chan *AnalyzeFastTask
+	rowCount        uint64
+	sampCursor      int32
+	sampTasks       []*AnalyzeFastTask
 	scanTasks       []*tikv.KeyLocation
+	collectors      []*statistics.SampleCollector
+}
+
+func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild *bool, err *error, sampTasks *[]*AnalyzeFastTask) {
+	defer func() {
+		e.wg.Done()
+		if *needRebuild == true {
+			for ok := true; ok; _, ok = <-e.sampLocs {
+				// Do nothing, just clear the channel.
+			}
+		}
+	}()
+	client := e.ctx.GetStore().(tikv.Storage).GetTiKVClient()
+	for {
+		loc, ok := <-e.sampLocs
+		if !ok {
+			return
+		}
+		req := &tikvrpc.Request{
+			Type: tikvrpc.CmdDebugGetRegionProperties,
+			DebugGetRegionProperties: &debugpb.GetRegionPropertiesRequest{
+				RegionId: loc.Region.GetID(),
+			},
+		}
+		var resp *tikvrpc.Response
+		var rpcCtx *tikv.RPCContext
+		rpcCtx, *err = e.cache.GetRPCContext(bo, loc.Region)
+		if *err != nil {
+			return
+		}
+		ctx := context.Background()
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			return
+		}
+		if resp.DebugGetRegionProperties == nil || len(resp.DebugGetRegionProperties.Props) == 0 {
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "mvcc.num_rows" || prop.Name == "num_rows" {
```

Comment:
what's the difference between "mvcc.num_rows" and "num_rows", it seems we have to build a sample task for both of them?

Reply:
The tikv uses "mvcc.num_rows" and the mocktikv uses the "num_rows".

Reply:
OK, I think it's better to add some comments for these two constants. BTW, once we find the property and build a sampTask for this region, It's better to break the `for` loop.

Reply:
Can we modify the mocktikv to unify them?

Reply:
OK.